out-of-order sync for share

aryairani opened the pull request:
closes #3038

This deserves a better writeup, and I should've done it during my self-review, but I didn't think of it until now and don't have time to go through it again right now —Arya

## Implementation notes

The PR adds a couple of sqlite tables: `temp_entity` and `temp_entity_missing_dependency`.
Sync entities (causals, namespaces, term/decl components, patches) are stored in the `temp_entity` table while they have missing dependencies (tracked in the `temp_entity_missing_dependency` table). Once the dependencies have been added, the entities are moved to the main tables, continuing with their unflushed dependents.

## Interesting/controversial decisions

Include anything that you thought twice about, debated, chose arbitrarily, etc. 
What could have been done differently, but wasn't? And why?

## Test coverage

Have you included tests (which could be a transcript) for this change, or is it somehow covered by existing tests? 

Would you recommend improving the test coverage (either as part of this PR or as a separate issue) or do you think it’s adequate?

## Loose ends

Link to related issues that address things you didn't get to. Stuff you encountered on the way and decided not to include in this PR.

Commit:
Serialization.getTemp{Namespace,Causal}Format

Commit:
Sync.{makeTempEntity,tempEntityType}

Commit:
partially broken stuff

Commit:
maybe done with download?

Commit:
tempToSyncTermComponent

mitchellwrosen committed:
get code compiling, bit of work on temp->main term

mitchellwrosen committed:
finish tempToSyncTermComponent

mitchellwrosen committed:
serialize in insertTempEntity

mitchellwrosen committed:
extract unison-util-base32hex package

mitchellwrosen committed:
use Base32Hex in unison-share-api

mitchellwrosen committed:
cleanup/refactoring

mitchellwrosen committed:
begin implementing entityToTempEntity

mitchellwrosen committed:
fill out most of entityToTempEntity

Commit:
added value hash to api type

mitchellwrosen committed:
use the real http update path

mitchellwrosen committed:
uploadEntities

mitchellwrosen committed:
move some code around and add docs

mitchellwrosen committed:
actually call downloadEntities

mitchellwrosen committed:
fill out pull

mitchellwrosen committed:
cleanup

Commit:
switch EntityLocation/2 around

Commit:
reorganize createSchema / migrateSchema3To4

Commit:
delete temp comments

Commit:
add fail responses to Sync.Types

Commit:
please review

- refactored saveCausal to handle parents as well
- moved Q.saveCausal in Ops.saveBranch to after looking up parent ids
- fleshing out Q.tryMoveTempEntityDependents (needs another name)
- implemented decompose/recomposeTerm/DeclFormat for Q.saveReadyEntity
- replaced sketchy usages of decompose/recomposeComponent with decompose/recomposeTerm/DeclFormat in Sync22.hs
  - needs audit
  - would hide decompose/recomposeComponent from export list, but we don't have one yet.

mitchellwrosen committed:
fix compiler error in migration

mitchellwrosen committed:
fix query

mitchellwrosen committed:
delete explicit imports of Unison.Prelude

mitchellwrosen commented on codebase2/codebase-sqlite/U/Codebase/Sqlite/Queries.hs:
```diff
@@ -486,17 +510,127 @@ updateObjectBlob oId bs = execute sql (oId, bs) where sql = [here|
+  dependents <- getMissingDependentsForTempEntity dependencyBase32
+  executeMany deleteTempDependents (dependents <&> (,dependencyBase32))
+  deleteTempEntity dependencyBase32
+  traverse_ moveTempEntityToMain =<< tempEntitiesWithNoMissingDependencies
```
Note from call: let's just return hash+blob+type_id from `tempEntitiesWithNoMissingDependencies`

mitchellwrosen commented on codebase2/codebase-sqlite/U/Codebase/Sqlite/ReadyEntity.hs:
```diff
@@ -0,0 +1,15 @@
+import qualified U.Codebase.Sqlite.Patch.Format as Patch
+import qualified U.Codebase.Sqlite.Term.Format as Term
+
+data ReadyEntity
```
Idea from call: unify with `TempEntity`

mitchellwrosen committed:
fill out the tempToSync* things

mitchellwrosen committed:
flush temp entities in saveObject

mitchellwrosen committed:
change some names

mitchellwrosen committed:
fill out tempToSyncEntity

mitchellwrosen committed:
fill out insertEntity

mitchellwrosen committed:
⅄ trunk → arya/ooo-sync

mitchellwrosen committed:
handle a couple missing cases

mitchellwrosen committed:
delete unused function

mitchellwrosen committed:
delete unused query

mitchellwrosen committed:
hit actual getCausalHashByPath endpoint

mitchellwrosen committed:
implement rowsModified

mitchellwrosen committed:
don't store hash jwts in temp_entity

mitchellwrosen committed:
unify ReadyEntity and TempEntity

mitchellwrosen committed:
take a connection rather than a transaction-running function, for incremental pulls

mitchellwrosen committed:
begin implementing expectEntity

mitchellwrosen committed:
move sqlite decoders into their own module

mitchellwrosen committed:
finish expectEntity

mitchellwrosen committed:
implement most of expectEntity in Share.Sync, except for handling local diff formats

mitchellwrosen committed:
add patch diff to share api types

mitchellwrosen committed:
tempEntityToEntity PatchDiff case

mitchellwrosen committed:
add namespace diff to share api types

mitchellwrosen committed:
some misc cleanup in sync code

mitchellwrosen committed:
⅄ trunk → arya/ooo-sync

mitchellwrosen committed:
add the codebase to LoopState

mitchellwrosen committed:
expose sqlite connection in codebase object

mitchellwrosen committed:
sketch out pushing to unison share from HandleInput

mitchellwrosen committed:
decode hash jwts

mitchellwrosen committed:
add unison share url to loop state env

mitchellwrosen committed:
write loadCausalHashAtPath

Commit:
Merge remote-tracking branch 'origin/arya/Wno-type-defaults' into arya/ooo-sync

mitchellwrosen committed:
more push.share work

mitchellwrosen committed:
push/pull work

Commit:
Merge remote-tracking branch 'origin/arya/ooo-sync' into arya/ooo-sync

mitchellwrosen committed:
add fast-forward path api

mitchellwrosen committed:
start implementing fast-forward push

mitchellwrosen committed:
sketch out pretty naive dag BFS algorithm

mitchellwrosen committed:
typo

Commit:
switch bfs path accumulator to support O(log n) append

mitchellwrosen committed:
replace force push with fast-forward push

mitchellwrosen committed:
⅄ trunk → arya/ooo-sync

mitchellwrosen committed:
⅄ trunk → arya/ooo-sync

mitchellwrosen committed:
push.create to share

mitchellwrosen committed:
PushError -> CheckAndSetPushError

mitchellwrosen committed:
organize share types module

mitchellwrosen committed:
delete unused TypedHash

Commit:
small refactor

mitchellwrosen committed:
initial work on making ReadRepo a sum type

mitchellwrosen committed:
hook up fast-forward push to dagbfs

Commit:
now compiles

Commit:
started adding ShareRepo constructors

mitchellwrosen committed:
inline Unison.Sync.HTTP into Unison.Share.Sync

Commit:
refactored RepoPath into a non-empty Path

Commit:
fix ShareRepo related missing case warnings

mitchellwrosen committed:
pull handleCreatePullRequest out to the top level

mitchellwrosen committed:
broken partial propagation of more structured read/write repo types

mitchellwrosen committed:
more busted work

mitchellwrosen committed:
fix a few compiler errors and make a better push type

mitchellwrosen committed:
pull git/share case

Commit:
importRemoteShareBranch and handleCreatePullRequest

ChrisPenner commented on codebase2/codebase-sqlite/U/Codebase/Sqlite/Branch/Format.hs:
```diff
@@ -35,17 +37,52 @@ data BranchFormat
-data SyncBranchFormat
-  = SyncFull BranchLocalIds ByteString
-  | SyncDiff BranchObjectId BranchLocalIds ByteString
+data SyncBranchFormat' parent text defn patch child
```
I think we'd planned to start using this naming scheme, right? Or have we appealed that decision? 😄 

```suggestion
data GSyncBranchFormat parent text defn patch child
```

aryairani replied:
Yeah, how did this end up with a `'` again?

mitchellwrosen replied:
Well, we did the tick to be consistent with the other nearby stuff.

aryairani replied:
Right, yeah. We had to choose between adding inconsistency, or ripping off the bandaid and trying to rename everything in this PR.

aryairani replied:
I'd rather merge before fixing, and do one PR later to rename everything along these lines that we want to rename. Does Github support hashtags? #techdebt

ChrisPenner commented on codebase2/codebase-sqlite/U/Codebase/Sqlite/Queries.hs:
```diff
@@ -171,8 +207,17 @@ import Unison.Sqlite
-  traverse_ (execute_ . fromString) $ List.splitOn ";" [hereFile|sql/create.sql|]
+createSchema = do
+  executeFile [hereFile|sql/create.sql|]
+  addTempEntityTables
```
Are we using these temp entity tables in UCM at all or is this an Enlil specific thing?

aryairani replied:
Yes, we use them on the client when pulling from Share.

ChrisPenner commented on codebase2/codebase-sqlite/sql/001-temp-entity-tables.sql:
```diff
@@ -0,0 +1,61 @@
+-- an entity proper (after storing all of its dependencies), we should always atomically delete the corresponding
+-- `temp_entity` row, if any.
+create table temp_entity (
+  hash text primary key not null,
```
Could you explain the reasoning to not use hash_id's here? This table should generally be small, so it's not a space constraint or anything, and it's probably easier to work with this way anyways, I'm just curious if it was thought about and what the decision was 😄

ChrisPenner commented on codebase2/codebase-sqlite/sql/001-temp-entity-tables.sql:
```diff
@@ -0,0 +1,61 @@
+  dependent text not null references temp_entity(hash),
+  dependency text not null,
+  dependencyJwt text not null,
+  unique (dependent, dependency)
```
This could likely be the primary key right?

ChrisPenner commented on codebase2/codebase-sqlite/sql/001-temp-entity-tables.sql:
```diff
@@ -0,0 +1,61 @@
+  dependencyJwt text not null,
+  unique (dependent, dependency)
+);
+create index temp_entity_missing_dependency_ix_dependent on temp_entity_missing_dependency (dependent);
```
I believe this is redundant given the `unique (dependent, dependency)` index.

ChrisPenner commented on codebase2/codebase-sqlite/U/Codebase/Sqlite/Queries.hs:
```diff
@@ -485,17 +568,279 @@ recordObjectRehash old new =
+tryMoveTempEntityDependents :: Base32Hex -> Transaction ()
+tryMoveTempEntityDependents dependencyBase32 = do
+  dependents <- getMissingDependentsForTempEntity dependencyBase32
+  executeMany deleteTempDependents (dependents <&> (,dependencyBase32))
```
Do we need to query for the missing dependents before deleting them? Doesn't look like we use the actual list of dependents, and AFAICT the combination of `getMissingDependentsForTempEntity` and `deleteTempDependents` are equivalent to:

```sql
DELETE FROM temp_entity_missing_dependency
  WHERE dependency = ?
```

which is a bit simpler and more efficient.

ChrisPenner commented on codebase2/codebase-sqlite/U/Codebase/Sqlite/Queries.hs:
```diff
@@ -485,17 +568,279 @@ recordObjectRehash old new =
+        TermFormat.SyncTerm . TermFormat.SyncLocallyIndexedComponent
+          <$> Lens.traverseOf (traverse . Lens._1) (bitraverse saveText expectObjectIdForHash32) terms
+
+syncToTempEntity :: SyncEntity -> Transaction TempEntity
```
I'm curious if this is something we want; if we have a SyncEntity, we know we have all the dependencies already because a SyncEntity uses object IDs, so when would we ever need to convert it into a TempEntity?

AFAICT it's used just because we already have a `tempEntityToEntity`, but it makes me wonder whether we might be doing more work than necessary by converting to the TempEntity, then converting from there to an Entity rather than doing the ideal conversion the first time.

If this is the easiest way that's fine, but maybe put a comment on it stating that the operation itself doesn't make much semantic sense and mention why it exists 😄

aryairani replied:
It does look like it's just used in that one spot. The SyncEntity and the TempEntity are different parameterizations of the same type, so it is natural to break this step of the conversion out, but it could be a local function instead of top-level if we wanted, or else a note.

ChrisPenner commented on codebase2/codebase-sqlite/U/Codebase/Sqlite/Queries.hs:
```diff
@@ -1047,3 +1405,78 @@ ancestorSql =
+--
+-- Preconditions:
+--   1. The entity does not already exist in "main" storage (`object` / `causal`)
+--   2. The entity does not already exist in `temp_entity`.
```
Does this need to be a pre-condition or could we just check and bail if it already exists?
Also, in the case that it does already exist, is there possibly something to be gained by overwriting the hashJWTs with fresher ones?

aryairani replied:
For 1) The way it's used in the one spot, `upsertEntitySomewhere`, we check if it's anywhere, and if it is, we bail. Otherwise we insert it into main storage, or into temp storage (this function).

For 2) It's currently written to skip if it's already in `temp_entity`, but would insert duplicates into `temp_entity_missing_dependency`, which seems not good.

Re: overwriting, sort of; back when we had expirations and stuff though, it would probably have been good to update, but you wouldn't be able to count on an update, which means it would have to be good enough without an update, in which case it doesn't matter.

ChrisPenner commented on lib/unison-prelude/src/Unison/Prelude.hs:
```diff
@@ -77,6 +79,14 @@ onNothing m may = maybe m pure may
 
+whenJust :: Applicative m => Maybe a -> (a -> m ()) -> m ()
+whenJust mx f =
+  maybe (pure ()) f mx
```
```
whenJust = for_
```

mitchellwrosen replied:
Yeah, but easier to read 😛

ChrisPenner replied:
Yeah I'd agree with that 👍🏼

ChrisPenner commented on parser-typechecker/src/Unison/Codebase/Type.hs:
```diff
@@ -174,7 +175,13 @@ data Codebase m v a = Codebase
+    -- At one time the codebase was meant to abstract over the storage layer, but it has been cumbersome. Now we prefer
+    -- to interact with SQLite directly, and so provide this temporary escape hatch, until we can eliminate this
+    -- interface entirely.
+    connection :: Sqlite.Connection
```
Is it feasible to provide `withConnection :: (Connection -> m r) -> m r` instead?
Exposing the connection directly allows the connection to leak out of the lifespan of the codebase itself, at which point the connection will have been closed.

ChrisPenner commented on unison-cli/src/Unison/Share/Sync.hs:
```diff
@@ -0,0 +1,764 @@
+    Share.UpdatePathHashMismatch mismatch -> pure (Left (CheckAndSetPushErrorHashMismatch mismatch))
+    Share.UpdatePathMissingDependencies (Share.NeedDependencies dependencies) -> do
+      -- Upload the causal and all of its dependencies.
+      uploadEntities httpClient unisonShareUrl conn (Share.RepoPath.repoName repoPath) dependencies >>= \case
```
Can we just call `checkAndSetPush` again after uploading dependencies? Or are you worried about the 'how did this happen' case causing an infinite loop?

mitchellwrosen replied:
What would be the purpose of that?

ChrisPenner replied:
Just code de-duplication, the cases are all the same 🤷🏼‍♂️

ChrisPenner commented on unison-cli/src/Unison/Share/Sync.hs:
```diff
@@ -0,0 +1,764 @@
+  | FoundGoal a
+
+-- FIXME: document
+dagbfs :: forall a m. Monad m => (a -> Bool) -> (a -> m [a]) -> a -> m (Maybe [a])
```
Not necessarily any better or worse, but I got nerd-sniped by BFS, so here's a version which uses logict (it includes the root in the result, but feel free to drop it if unneeded);

```haskell
dagbfs :: forall a m. (MonadIO m, Show a) => (a -> Bool) -> (a -> m [a]) -> a -> m (Maybe [a])
dagbfs goal getChildren root =
  fmap listToMaybe . observeManyT 1 . flip runReaderT [] $ (bfs root >>= guardMaybe)
  where
    guardMaybe Nothing = empty
    guardMaybe (Just a) = pure a
    bfs :: a -> ReaderT [a] (LogicT m) (Maybe [a])
    bfs child = local (child :) $ do
      liftIO $ print child
      -- For a proper breadth-first traversal we emit failures too.
      let nodeResult =
            if goal child
              then Just <$> ask
              else pure Nothing
      nodeResult <|> do
        children <- lift . lift $ getChildren child
        interleaveAll (bfs <$> children)

-- |
-- >>> observeAll $ interleaveAll [pure @Logic 1, pure 2 <|> pure 3 <|> pure 4, pure 5, pure 6 <|> pure 7]
-- [1,2,5,6,3,7,4]
interleaveAll :: forall f a. MonadLogic f => [f a] -> f a
interleaveAll [] = empty
interleaveAll xs = do
  go xs []
  where
    go :: [f a] -> [f a] -> f a
    go [] ms = interleaveAll ms
    go (x : rest) ms = do
      msplit x >>= \case
        Nothing -> go rest ms
      -- use a DList probably
        Just (a, m) -> pure a <|> go rest (ms ++ [m])
```

mitchellwrosen committed:
fill in call to push to unison share

Commit:
filled in a lot of things?

Commit:
started fleshing out OutputMessages

Commit:
working on OutputMessages

mitchellwrosen committed:
prettyRemoteNamespace -> prettyReadRemoteNamespace

mitchellwrosen committed:
more pretty-printing work

mitchellwrosen committed:
couple more output messages

Commit:
implementing the new git and share remote syntaxes, and maybe other stuff

Commit:
Merge branch 'ooo-sync-read-repo-sum-type' into arya/ooo-sync

Commit:
plumb new repo types

mitchellwrosen committed:
⅄ ooo-sync-read-repo-sum-type → arya/ooo-sync

ChrisPenner commented on lib/unison-prelude/src/Unison/Prelude.hs:
```diff
@@ -165,3 +166,7 @@ reportBug bugId msg =
     ]
+
+{-# WARNING wundefined "You left this wundefined." #-}
+wundefined :: a
```
Would probably still be nice to have a call-stack here.

```suggestion
wundefined :: HasCallStack => a
```

mitchellwrosen committed:
⅄ trunk → arya/ooo-sync

Commit:
some new operations for enlil

Commit:
Merge branch 'arya/ooo-sync' of github.com:unisonweb/unison into arya/ooo-sync

mitchellwrosen committed:
get syncCausalHash building

mitchellwrosen committed:
hpack tweak

mitchellwrosen committed:
saveDbBranch

ChrisPenner committed:
Expose Codebase connection

ChrisPenner committed:
Fix redundant import

Commit:
reexport Data.Tuple.Only, move entityExists, update Sync error types

Commit:
export HashMismatchForEntity and InvalidParentage

Commit:
add FastForwardPathrequest.expectedHash field

Commit:
fix json instances

mitchellwrosen committed:
tweak fast-forward path to align with server's expectations

ChrisPenner committed:
Adds ToJWT/FromJWT instances for HashJWT

ChrisPenner committed:
Merge remote-tracking branch 'origin/arya/ooo-sync' into arya/ooo-sync

ChrisPenner committed:
Add missing error handling

mitchellwrosen committed:
use BranchHash / PatchHash where appropriate

ChrisPenner committed:
Expose implementations for use in Share server

ChrisPenner committed:
Merge remote-tracking branch 'origin/arya/ooo-sync' into arya/ooo-sync

ChrisPenner committed:
Remove entity type from hashjwt for now.

mitchellwrosen committed:
change upsertEntitySomewhere return type to be more clear

ChrisPenner committed:
Merge remote-tracking branch 'origin/arya/ooo-sync' into arya/ooo-sync

Commit:
download entities

ChrisPenner committed:
Add traversals for Entity types

Commit:
json instances for DownloadEntitiesResponse

ChrisPenner committed:
Merge remote-tracking branch 'origin/arya/ooo-sync' into arya/ooo-sync

Commit:
Merge branch 'arya/ooo-sync' of github.com:unisonweb/unison into arya/ooo-sync

Commit:
comment out `DownloadEntities`

aryairani committed:
Update lib/unison-prelude/src/Unison/Prelude.hs

Co-authored-by: Chris Penner <christopher.penner@gmail.com>

mitchellwrosen committed:
fix bug in entityLocation

mitchellwrosen committed:
generalize entityToTempEntity

mitchellwrosen committed:
move causalHashToShareHash to common

Commit:
Merge remote-tracking branch 'origin/trunk' into arya/ooo-sync

# Conflicts:
#	codebase2/codebase-sqlite/U/Codebase/Sqlite/Queries.hs
#	parser-typechecker/package.yaml
#	parser-typechecker/src/Unison/Codebase.hs
#	parser-typechecker/src/Unison/Codebase/SqliteCodebase.hs
#	parser-typechecker/src/Unison/Codebase/SqliteCodebase/Migrations/MigrateSchema3To4.hs
#	unison-cli/src/Unison/Codebase/Editor/Command.hs
#	unison-cli/src/Unison/Codebase/Editor/HandleCommand.hs
#	unison-cli/src/Unison/Codebase/Editor/Output.hs
#	unison-cli/src/Unison/Codebase/Editor/VersionParser.hs
#	unison-cli/src/Unison/CommandLine/OutputMessages.hs

Commit:
fix up merge compilation errors

and formatting differences

Commit:
fix up tests

Commit:
Merge remote-tracking branch 'origin/trunk' into arya/ooo-sync

Commit:
fixing up some sync bugs

Commit:
update tryMoveTempEntityDependents

Commit:
fix tryMoveTempEntityDependents differently

mitchellwrosen committed:
wrap a couple output messages (for spaces)

Commit:
flushIfReadyToFlush

Commit:
Merge branch 'arya/ooo-sync' of github.com:unisonweb/unison into arya/ooo-sync

ChrisPenner committed:
Fix incorrect ToJWT/FromJWT instances for HashJWT

Commit:
tweak fixes

ChrisPenner committed:
Extract update names index logic for easier use with sync algorithms

ChrisPenner committed:
Merge remote-tracking branch 'origin/arya/ooo-sync' into arya/ooo-sync

Commit:
add when/Paths_ stuff to every package.yaml target

didn't put it into another PR, I'm sorry. But we can cherry-pick this out in a separate PR

ChrisPenner committed:
Allow passing an access token explicitly

This lets us use transcripts for testing codeservers

ChrisPenner committed:
Remove unused sync http module

ChrisPenner committed:
Provide credential manager and Authenticated client in transcripts

Can't think of a good reason not to allow it

ChrisPenner committed:
docs

ChrisPenner committed:
Maybe fix windows failures?

Commit:
add comments and delete dead code

Commit:
fix branch TempEntity serialization mismatch

Commit:
delete dead utility functions

Commit:
ormolu

Commit:
hpack

ChrisPenner committed:
Update input patterns and unisonConfig for Share paths. (#3080)

* Clean up docs and parsers for read/write remote namespaces for share

* Revive deprecated git parser so old unisonConfig's still work

* Don't hide git errors

* Fix CodeserverURI -> URI conversion

* Reflect the provided port in CodeserverURI

* Focus messaging around primary action

ChrisPenner committed:
Commit cabal changes

ChrisPenner committed:
Merge remote-tracking branch 'origin/arya/ooo-sync' into arya/ooo-sync

Commit:
delete hie.yaml

Commit:
Merge branch 'arya/ooo-sync' of github.com:unisonweb/unison into arya/ooo-sync

mitchellwrosen committed:
begin implementing push/pull progress reporting

mitchellwrosen committed:
leave upload/download counts up after push/pull are complete

mitchellwrosen committed:
remove optimized Int->Text conversion since it wasn't impressively faster

mitchellwrosen committed:
⅄ trunk → arya/ooo-sync

mitchellwrosen committed:
adjust contrib/cabal.project

mitchellwrosen committed:
add back hie.yaml

mitchellwrosen committed:
Add U.Util.Hash32

mitchellwrosen committed:
replace Share.Hash with U.Util.Hash32

aryairani committed:
-fno-warn-orphans in orphans modules

aryairani committed:
turn on some timing stuff temporarily

aryairani committed:
add `seen` check to elaborateHashes

aryairani committed:
prepare to be a submodule called `unison`

aryairani committed:
fix sloppy checkin c1489998

mitchellwrosen committed:
pull some console printing out to a helper

mitchellwrosen committed:
⅄ arya/ooo-sync → ooo-sync-push-pull-progress

aryairani committed:
recursive query for server-side elaborateHashes

mitchellwrosen committed:
slightly adjust placement of upload callback in push to be more correct

mitchellwrosen committed:
chunk share uploads (currently 500 entities / request)

mergify[bot] committed:
Merge pull request #3097 from unisonweb/ooo-sync-push-pull-progress

Share push/pull: show some indication of progress

aryairani committed:
recursive query for client-side elaborateHashes (not yet testable)

aryairani committed:
Merge branch 'arya/ooo-sync' of github.com:unisonweb/unison into arya/ooo-sync

ChrisPenner committed:
Merge branch 'arya/ooo-sync' into cp/allow-apis-in-transcripts-merged

ChrisPenner committed:
Implement token refresh (#3095)

mergify[bot] committed:
Merge pull request #3092 from unisonweb/cp/allow-apis-in-transcripts-merged

Allow using credentials for codeservers in transcripts

aryairani committed:
entity location queries

aryairani committed:
Merge branch 'arya/ooo-sync' of github.com:unisonweb/unison into arya/ooo-sync

mitchellwrosen committed:
delete duplicated EntityLocation/entityLocation and deal with the
descrepancy

ChrisPenner committed:
you are no longer bad

aryairani committed:
fix some bad sql syntax

aryairani committed:
Merge branch 'arya/ooo-sync' of github.com:unisonweb/unison into arya/ooo-sync

aryairani committed:
draft - support alternate pull implementation

tstat committed:
convert2

mitchellwrosen committed:
Merge pull request #3102 from unisonweb/ooo-sync-smaller-pushes

Limit batch size of pushes to share

aryairani committed:
update pull' stuff

aryairani committed:
Merge branch 'arya/ooo-sync' of github.com:unisonweb/unison into arya/ooo-sync

mitchellwrosen committed:
work towards replacing old pull with new better pull

mitchellwrosen committed:
more minor cleanup

mitchellwrosen committed:
on push, show number of entities enqueued to push too

mitchellwrosen committed:
drop push entity size from 500 to 50 (for now)

mitchellwrosen committed:
slightly faster expectObjectIdForHash32

mitchellwrosen committed:
add saveTexts, use it in a couple places

aryairani committed:
fix warnings

mitchellwrosen committed:
disable client timeouts for sync endpoints

mitchellwrosen committed:
don't upload or fast-forward push at all if local == remote

mitchellwrosen committed:
remove unnecessary delete

mitchellwrosen committed:
make part of trying to flush temp entities more efficient

mitchellwrosen committed:
bump direct-sqlite dep

mitchellwrosen committed:
minor sql improvements

mitchellwrosen committed:
collate nocase on where clauses involving hash.base32

mitchellwrosen committed:
delete unused import

mitchellwrosen committed:
⅄ trunk → arya/ooo-sync

ChrisPenner committed:
Use Output for credential failures (#3106)

aryairani committed:
move hashjwt decoding up

aryairani committed:
Merge branch 'arya/ooo-sync' of github.com:unisonweb/unison into arya/ooo-sync

# Conflicts:
#	codebase2/codebase-sqlite/U/Codebase/Sqlite/Queries.hs

aryairani committed:
cleanup unused import

aryairani committed:
tweak sqlite package version comment

aryairani commented on codebase2/codebase-sqlite/U/Codebase/Sqlite/Branch/Format.hs:
```diff
@@ -35,17 +37,52 @@ data BranchFormat
-    branchChildLookup :: Vector (BranchObjectId, CausalHashId)
+type BranchLocalIds = BranchLocalIds' TextId ObjectId PatchObjectId (BranchObjectId, CausalHashId)
+
+-- temp_entity
+--  branch #foo
+--
+-- temp_entity_missing_dependency
+--   #foo depends on causal #bar
+--   #foo depends on namespace #baz
+--
+-- 1. store causal #bar, go to flush dependencies like normal
+-- 2. ... oh this case is different than the others - we don't want to delete that row
+
+----
+-- can't simply treat causal's value hash as a mandatory dependency because we can't be sure
+-- that the causal doesn't already exist in the target codebase without the value.
+-- it probably does exist together with the value (though we found cases in the past where it didn't
+-- due to race conditions, but we fixed that and added transactions and it shouldn't happen again?),
+-- but it's not enforced at the schema level.
+-- to enforce it at the schema level, we'd have to do something like store namespace_object_id instead
+-- of value_hash in causal, which would require a db migration for a thing we don't necessarily even want
+-- long term.
+-- so, we can't simply "require" the value hash as a dependency of the causals and expect things to work smoothly
+-- without relying on prayer.
+--
+-- temp_entity
+--  branch #foo
+--  causal #bar
+
+-- temp_entity_missing_dependency
+--   #foo depends on causal #bar
+--   #bar depends on namespace #baz
+--
+
```
delete

aryairani replied:
```suggestion
```

tstat committed:
Add unison-codebase-sqlite-hashing-v2 package

unison-codebase-sqlite is now parameterized by the hashing functions

aryairani committed:
indentation

tstat committed:
merge trunk

ChrisPenner committed:
Skip all the extra claims on HashJWTs (#3107)

* Simplify hashjwt claims

* update json instances

* Add jwt type to the hashjwt

Co-authored-by: Mitchell Rosen <mitchellwrosen@gmail.com>

aryairani committed:
Merge branch 'arya/ooo-sync' of github.com:unisonweb/unison into arya/ooo-sync

ChrisPenner committed:
Better messaging for transport errors during sync

ChrisPenner committed:
Use codeserver url in messages

ChrisPenner committed:
Formatting fixes

mitchellwrosen committed:
make pull from share download in smaller chunks

mitchellwrosen committed:
do only one transaction in `downloadEntities`

mitchellwrosen committed:
move elaborateHashesServer to enlil

mitchellwrosen committed:
rename elaborateHashesClient to elaborateHashes

mitchellwrosen committed:
Merge pull request #3115 from unisonweb/22-06-15-ooo-sync-chunked-pull

make pull from share download in smaller chunks

mitchellwrosen committed:
expose a saveTempEntityInMain instead of saveSyncEntity

mitchellwrosen committed:
don't export tempToSyncEntity now that there are no external callers anymore

mitchellwrosen committed:
do some documentation and cleanup

aryairani committed:
Merge branch 'arya/ooo-sync' of github.com:unisonweb/unison into arya/ooo-sync

aryairani committed:
unison-core1 doesn't need util

aryairani committed:
Merge remote-tracking branch 'origin/trunk' into arya/ooo-sync

# Conflicts:
#	hie.yaml
#	parser-typechecker/package.yaml
#	unison-hashing-v2/package.yaml
#	unison-hashing-v2/unison-hashing-v2.cabal

tstat committed:
Fix error from arya/ooo-sync rebase

ChrisPenner committed:
Update unison-cli/src/Unison/CommandLine/OutputMessages.hs

Co-authored-by: Arya Irani <538571+aryairani@users.noreply.github.com>

ChrisPenner committed:
Fix bad syntax on PR suggestion

tstat committed:
remove all effect vars on all funcs in HashHandle

aryairani commented on codebase2/codebase-sqlite/U/Codebase/Sqlite/Queries.hs:
```diff
@@ -491,17 +629,305 @@ recordObjectRehash old new =
+
+{- ORMOLU_DISABLE -}
+
+syncToTempEntity :: SyncEntity -> Transaction TempEntity
```
```suggestion
-- | looking up all of the text and hashes is the first step of converting a SyncEntity to a Share.Entity
syncToTempEntity :: SyncEntity -> Transaction TempEntity
```

mergify[bot] committed:
Merge pull request #3114 from unisonweb/cp/transport-errors

Better messaging for Transport Errors when speaking to a code server

aryairani commented on codebase2/codebase-sqlite/U/Codebase/Sqlite/Serialization.hs:
```diff
@@ -633,7 +646,25 @@ getBranchLocalIds =
     <*> getVector (getPair getVarInt getVarInt)
 
-decomposeComponent :: MonadGet m => m [(LocalIds, BS.ByteString)]
+decomposeTermFormat :: MonadGet m => m TermFormat.SyncTermFormat
+decomposeTermFormat =
+  getWord8 >>= \case
+    0 ->
+      TermFormat.SyncTerm
+        . TermFormat.SyncLocallyIndexedComponent
+        <$> decomposeComponent
+    tag -> error $ "todo: unknown term format tag " ++ show tag
```
fixing up a long-standing weirdness

aryairani commented on codebase2/util/src/U/Util/Timing.hs:
```diff
@@ -10,7 +10,7 @@ import System.IO.Unsafe (unsafePerformIO)
 
 enabled :: Bool
-enabled = False
+enabled = True
```
```suggestion
enabled = False
```

aryairani commented on parser-typechecker/package.yaml:
```diff
@@ -13,119 +13,120 @@ when:
     ghc-options: -funbox-strict-fields -O2
 
+dependencies:
+  - ListLike
+  - NanoID
```
I guess these ended up here because capital letters?

aryairani commented on unison-cli/src/Unison/Share/Codeserver.hs:
```diff
@@ -0,0 +1,33 @@
+defaultCodeserver :: CodeserverURI
+defaultCodeserver = unsafePerformIO $ do
+  lookupEnv "UNISON_SHARE_HOST" <&> \case
+    -- TODO: swap to production share before release.
```
Is this still a todo @ChrisPenner, or can we just call it share-next?

aryairani commented on unison-cli/src/Unison/Codebase/Editor/UriParser.hs:
```diff
@@ -39,30 +55,157 @@ type P = P.Parsec Void Text.Text
-repoPath = P.label "generic git repo" $ do
-  protocol <- parseProtocol
-  treeish <- P.optional treeishSuffix
+repoPath =
+  P.label "generic repo" $
+    fmap ReadRemoteNamespaceGit readGitRemoteNamespace
+      <|> fmap ReadRemoteNamespaceShare readShareRemoteNamespace
```
I was thinking we had a fallback to unannotated git repos after share, but I guess not, and maybe it would be dead code anyway

aryairani commented on unison-share-api/src/Unison/Sync/Common.hs:
```diff
@@ -0,0 +1,204 @@
+        }
+
+    mungeNamespaceLocalIds ::
+      [Text] ->
```
I forget where these lists come from, and should we use Vector from the start?

aryairani commented on unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:
```diff
@@ -1697,51 +1778,75 @@ handlePushRemoteBranch ::
-    (repo, remotePath) <- maybe (resolveConfiguredGitUrl Push path) pure mayRepo
-    lift (doPushRemoteBranch repo path syncMode (Just (remotePath, pushBehavior)))
+handlePushRemoteBranch mayRepo path pushBehavior syncMode =
+  time
+    "handlePushRemoteBranch"
```
Could remove this, or just disable timing, or set to an environment variable.

ChrisPenner committed:
Update Name lookup index in half the time (#3124)

* Implement name lookup on shallow branches

ChrisPenner committed:
Redirect back to share URI after successful login (#3128)

ChrisPenner committed:
Ensure reversed names are actually reversed.

ChrisPenner committed:
Fix mis-typed sqlite columns in name lookup index

ChrisPenner committed:
Add ensureAuthenticated combinator (#3127)

tstat committed:
Update codebase2/codebase-sqlite/U/Codebase/Sqlite/HashHandle.hs

Co-authored-by: Mitchell Rosen <mitchellwrosen@gmail.com>

tstat committed:
Update codebase2/codebase-sqlite/U/Codebase/Sqlite/Queries.hs

Co-authored-by: Mitchell Rosen <mitchellwrosen@gmail.com>

tstat committed:
no RecordWildCards

aryairani committed:
Merge pull request #3125 from unisonweb/travis/update-indices

Update type and term indices after push and pull

aryairani committed:
Update codebase2/codebase-sqlite/U/Codebase/Sqlite/Queries.hs
